controller_manager/src/controller_manager.cpp

```diff
@@ -18,135 +18,50 @@
-  bool is_absolute()
-  {
-    return path_[0] == '/';
-  }
-
-private:
-  std::string path_;
-};
-}  // namespace fs
-#else
-# include <experimental/filesystem>
-namespace fs = std::experimental::filesystem;
-#endif
 
 namespace controller_manager
 {
-
-static constexpr const char * resource_index = "ros_controllers";
-
-namespace
+ControllerManager::ControllerManager(
+  std::shared_ptr<hardware_interface::RobotHardware> hw,
+  std::shared_ptr<rclcpp::executor::Executor> executor, const std::string & manager_node_name)
+: rclcpp::Node(manager_node_name),
+  hw_(hw),
+  executor_(executor),
+  // add pluginlib loader by default
+  loaders_({std::make_shared<ControllerLoaderPluginlib>()})
 {
+}
 
-std::string
-parse_library_path(
-  const std::string & package_name,
-  const std::string & class_name,
-  const rclcpp::Logger logger)  // should be const ref when possible
+std::shared_ptr<controller_interface::ControllerInterface>
+ControllerManager::load_controller(
+  const std::string & controller_name,
+  const std::string & controller_type)
 {
-  // get node plugin resource from package
-  std::string content;
-  std::string base_path;
-  if (!ament_index_cpp::get_resource(resource_index, package_name, content, &base_path)) {
-    auto error_msg = std::string("unable to load resource for package ") + package_name;
-    RCLCPP_ERROR(logger, error_msg);
-    throw std::runtime_error(error_msg);
-  }
+  RCUTILS_LOG_INFO("Loading controller '%s'\n", controller_name.c_str());
 
-  auto allocator = rcutils_get_default_allocator();
-  rcutils_string_array_t controller_array = rcutils_get_zero_initialized_string_array();
-  rcutils_split(content.c_str(), '\n', allocator, &controller_array);
-  if (controller_array.size == 0) {
-    auto error_msg = std::string("no ros controllers found in package " + package_name);
-    RCLCPP_ERROR(logger, error_msg);
-    throw std::runtime_error(error_msg);
-  }
+  std::shared_ptr<controller_interface::ControllerInterface> controller(nullptr);
 
-  std::string library_path = "";
-  bool controller_is_available = false;
-  for (auto i = 0u; i < controller_array.size; ++i) {
-    rcutils_string_array_t controller_details = rcutils_get_zero_initialized_string_array();
-    rcutils_split(controller_array.data[i], ';', allocator, &controller_details);
-    if (controller_details.size != 2) {
-      auto error_msg = std::string("package resource content has wrong format") +
-        " - should be <class_name>;<library_path> but is " + controller_array.data[i];
-      RCLCPP_ERROR(logger, error_msg);
-      throw std::runtime_error(error_msg);
-    }
-
-    if (strcmp(controller_details.data[0], class_name.c_str()) == 0) {
-      library_path = controller_details.data[1];
-      if (!fs::path(library_path).is_absolute()) {
-        library_path = base_path + "/" + controller_details.data[1];
-      }
-      controller_is_available = true;
+  for (auto loader = loaders_.cbegin(); loader != loaders_.cend(); ++loader) {
```

Comment:
consider `std::find_if` with the appropriate lambda.

Reply:
Done!